### DISCUSSION TITLE
Adapt predictor matrix: impute variable only for specific years.

### DISCUSSION THREAD

**Discussion by MLindevd** [amices/mice#568] [Q&A]
I have a dataset for the years 2017-2022. I need to create a variable, which is based on two variables: 1 is collected in the years 2017-2020, and the other during the years 2021-2022. I would define the variable as the following: 

df <- df %>%
  mutate(radicaal = as.factor(case_when(year< 2021 & resec == 0 ~ "Radical",
                                        year< 2021 & resec == 1 | resec == 2 ~ "Not radical",
                                        year>= 2021 & microrad == 1 ~ "Radical",
                                        year>= 2021 & microrad == 2 ~ "Not radical")))

These variables were have NAs for the years in which they were not collected, but there are also NAs during the years I want to use them. 

I only want to impute the missing values for "resec"  in the years 2017-2021, and "microrad" only in the years 2021-2022. Is there any way that I can  change the predictor matrix? I tried the code below, but it doesn't work.. 

#### Imputation Set-Up ####
imp0 <- mice(dfslok, maxit = 0)

##Imputation method
meth <- imp0$method

#predictorMatrix
pred <- imp0$predictorMatrix
pred[,"resec"] <- dfslok$year>= 2017 & dfslok$year<= 2020 
pred[,"microrad"] <- dfslok$year>= 2021 & dfslok$year<= 2022 

Thanks in advance!

**Comment by stefvanbuuren** [comment]
It sounds as if it were easier to do if you organise the data in wide format, with different years in different columns.

**Reply by MLindevd** [reply]
Hi Stef, Thanks for your soon reply. However, it is not longitudinal data, but just multiple years of data and during these years some variable collections changed (like the "resec"/"microrad". Can I also do this by changing the predictormatrix?

**Reply by stefvanbuuren** [reply]
Then I fear I don't know what your data look like. I would suggest making a small reprex with a subset of the data.

**Comment by thomvolker** [comment]
I don't think this is possible by using the `predictorMatrix` argument, but you can specify which cells you want to impute though the `where`-argument. The `where` argument takes as input a logical matrix of the same dimensions as the input data, with the value `TRUE` for every cell that you want to impute. Setting all cells that you don't want to impute to `FALSE` makes sure that you won't impute the values that should not be imputed.
The downside is that if observations have missings on other variables as well, you cannot use the variables in which you keep the missings in the imputation models for the other variables in which you want to impute the missings.
